slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -329,6 +330,123 @@ void SLAMServiceImpl::SetUpMapBuilder() {
+            LOG(ERROR) << "Error creating pcd map: " << e.what();
+            throw std::runtime_error(e.what());
+        }
+    }
+
+    auto painted_surface = painted_slices->surface.get();
+    int width = cairo_image_surface_get_width(painted_surface);
+    int height = cairo_image_surface_get_height(painted_surface);
+    // all pixels from the painted surface in RGBA format
+    auto data = cairo_image_surface_get_data(painted_surface);
+
+    // total number of bytes in image (4 bytes per pixel)
+    int size_data = width * height * 4;
+
+    // each pixel contains 4 bytes of information in RGBA format
+    // data_vect[i + 0] is the R channel
+    // data_vect[i + 1] is the B channel
+    // data_vect[i + 2] is the G channel
+    // data_vect[i + 3] is the A channel
+    std::vector<unsigned char> data_vect(data, data + size_data);
+
+    int num_points = 0;
+
+    // Sample the image based off number of pixels. Output is number pixels to
+    // skip. Ideally should be between 5 and 15, but depends on the resolution
+    // of the original image
+    int skip_count = (size_data / (float)maximumGRPCByteLimit) * samplingFactor;
+    if (skip_count == 0) {
+        skip_count = 1;
+    }
+
+    std::string data_buffer;
+
+    // Loop to sample data and reduce resolution. Increments multiplied
+    // by 4 to represent 4 bytes per pixel
+    for (int i = 0; i < size_data; i += skip_count * 4) {
+        // skip pixels that are not in our map(black/past walls)
+        // this check represents [102,102,102]
+        if ((data_vect[i + 0] == defaultCairosEmptyPaintedSlice) &&
+            (data_vect[i + 1] == defaultCairosEmptyPaintedSlice) &&
+            (data_vect[i + 2] == defaultCairosEmptyPaintedSlice))
+            continue;
+
+        // Determine probability based off color pixel
+        int prob = viam::ViamColorToProbability((int)data_vect[i + 2]);
+        if (prob == 0) continue;
+
+        num_points++;
+
+        int pixel_index = i / 4;
+        int pixel_x = pixel_index % width;
+        int pixel_y = pixel_index / width;
+
+        // convert pixel location to pointcloud point in meters
+        float x_pos = (pixel_x - painted_slices->origin.x()) * kPixelSize;
+        // Y is inverted to match output from getPosition()
+        float y_pos = -(pixel_y - painted_slices->origin.y()) * kPixelSize;
+        // 2D SLAM so Z is set to 0
+        float z_pos = 0;
+
+        // rotating coordinates to match slam service expectation(XZ plane)
+        viam::utils::writeFloatToBufferInBytes(data_buffer, x_pos);
+        viam::utils::writeFloatToBufferInBytes(data_buffer, z_pos);
+        viam::utils::writeFloatToBufferInBytes(data_buffer, y_pos);
+        viam::utils::writeIntToBufferInBytes(data_buffer, prob);
+    }
+
+    // Write our PCD file, which is written as a binary.
+    pointcloud = viam::utils::pcdHeader(num_points, true);
+
+    // writes data buffer to the pointcloud string
+    pointcloud += data_buffer;
```

Comment:
@tessavitabile do we need to worry about the data from `data_buffer` is being copied into `pointcloud`  (meaning if `data_buffer` is 32MB, 64 MB will be consumed until the end of this function)?

From https://cplusplus.com/reference/string/string/operator+=/ it appears as if `data_buffer` will be copied onto `pointcloud`.

Do we need to write this such that the data from `data_buffer` is moved into `pointcloud` (thus preventing any unnecessary copying & ensuring that if `data_buffer` is 32MB, only 32MB are consumed within this function?

If so, how does one append to a string by moving (not copying) the other string into it?

Reply:
I don't think it's possible to move a string into another string. If we're really worried about memory consumption, we shouldn't use data_buffer at all, and instead iterate through once to count the number of points that will be in the pcd, and then iterate through again to add all the points.

Reply:
Whether we should be concerned, I'm not sure. It would be better to avoid the copy, but this memory might be small compared to overall memory used by cartographer, so it might not be worth worrying about.

Reply:
Cool, sounds good.